core/common/src/YearMonth.kt

```diff
@@ -0,0 +1,396 @@
+ * See sample 3.
+ *
+ * Additionally, there are several `kotlinx-serialization` serializers for [YearMonth]:
+ * - [YearMonthIso8601Serializer] for the ISO 8601 extended format.
+ * - [YearMonthComponentSerializer] for an object with components.
+ *
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.constructorFunctionMonthNumber
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.simpleParsingAndFormatting
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.customFormat
+ */
+@Serializable(with = YearMonthIso8601Serializer::class)
+public class YearMonth
+/**
+ * Constructs a [YearMonth] instance from the given year-month components.
+ *
+ * The [month] component is 1-based.
+ *
+ * The supported ranges of components:
+ * - [year] the range is platform-dependent, but at least is enough to represent year-months of all instants between
+ *          [Instant.DISTANT_PAST] and [Instant.DISTANT_FUTURE] in any time zone.
+ * - [month] `1..12`
+ *
+ * @throws IllegalArgumentException if any parameter is out of range.
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.constructorFunctionMonthNumber
+ */
+public constructor(year: Int, month: Int) : Comparable<YearMonth> {
+    /**
+     * Returns the year component of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.year
+     */
+    public val year: Int = year
+
+    /**
+     * Returns the number-of-the-month (1..12) component of the year-month.
+     *
+     * Shortcut for `month.number`.
+     */
+    internal val monthNumber: Int = month
+
+    init {
+        require(month in 1..12) { "Month must be in 1..12, but was $month" }
+        require(year in LocalDate.MIN.year..LocalDate.MAX.year) {
+            "Year $year is out of range: ${LocalDate.MIN.year}..${LocalDate.MAX.year}"
+        }
+    }
+
+    /**
+     * Returns the month ([Month]) component of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.month
+     */
+    public val month: Month get() = Month(monthNumber)
+
+    /**
+     * Returns the first day of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.firstAndLastDay
+     */
+    public val firstDay: LocalDate get() = onDay(1)
+
+    /**
+     * Returns the last day of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.firstAndLastDay
+     */
+    public val lastDay: LocalDate get() = onDay(numberOfDays)
+
+    /**
+     * Returns the number of days in the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.numberOfDays
+     */
+    public val numberOfDays: Int get() = monthNumber.monthLength(isLeapYear(year))
+
+    /**
+     * Returns the range of days in the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.days
+     */
+    // val days: LocalDateRange get() = firstDay..lastDay // no ranges yet
```

Comment:
What is the reason for commenting this out? Seems useful to me.

Reply:
The reason is that `LocalDateRange` is not in `master`. See the description of the pull request.

Reply:
While LocalDateRange may not be in master, you can for now implement this as:
```
val days: ClosedRange<LocalDate> get() = firstDay..lastDay
```

Reply:
Actually, as of 2 days ago it looks like LocalDateRange is in master